<a href="https://colab.research.google.com/github/sebastianlh1588-beep/Data-Science-Portafolio---Mateo-L-pez-Huerta/blob/main/PAC_2027_UAI_ETAPA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# PROYECTO - ELABORACION DEL PLAN ANUAL DE CONTROL 2027
**Institución:** Empresa Publica de Telecomunicaciones y Agua Potable de Cuenca

**Departamento:** Unidad de Auditoria Intera de ETAPA EP

**Autor:** Mateo López Huerta

**Base de Datos:** Procesos de Contratación de ETAPA 2022 AL Sep/2026

**Fuente:** SERCOP

# **Contexto:**
El articulo 16.- Coordinacion de Laborales de la LOCGE y el articulo 10 del Acuerdo 004-CG-2021, establece que el Plan Anual de Control debe ser presentado hasta el 30 de septiembre de cada año, asi mismo. Mediante Acuerdo 038-CG-2026 el Contralor General Estado expide el Instructivo para la elaboracion del PAC 2027. En este sentido, el objetivo del proyecto es establecer lo procesos de contratacion mas riesgosos con el fin de incluirlos en el PAC 2027.
# Problematica:
1. Enteder cuales son los procesos de contratacion que son mas riesgosos.
2. Clasificación: Clasificar los procesos de contratacion mas riesgosos, estableciendo un score de priorización.  



# 1. Importacion de Librerias
En este primer paso se preparan las bases del proyecto mediante la carga de herramientas esenciales: Pandas y NumPy para estructurar y manipular el conjunto de datos de los procesos de contratacion de ETAPA EP del 2022 al 2026.

In [ ]:
# ==============================================================================
# PASO 1: ENTENDIMIENTO Y RECOLECCIÓN DE DATOS
# ==============================================================================
# Importamos las librerías fundamentales para manipulación, gráficos y modelado
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()


# Descripción:
Este bloque inicial de código prepara el entorno de trabajo configurando los cimientos del proyecto; primero, importa las librerías esenciales para la ciencia de datos, utilizando Pandas para la manipulación de tablas y NumPy para el soporte de operaciones matemáticas avanzadas. Ademas se utilizará la libreria sklearn importando minmaxscaler para nomarlizar la escala y que las variables sean comparables en la metodologia utilizada

In [ ]:
# Cargamos el dataset desde el archivo Excel adjunto utilizando read_excel
df = pd.read_excel('/content/drive/MyDrive/Contraloria UAI ETAPA EP/PAC 2027/PROCESOS DE CONTRATACION CONSOLI 2022 AL 2026 etapa ep.xlsx')

Aquí se utiliza la función read_excel de la librería Pandas para leer el archivo que contiene los datos de los procesos de contratación de ETAPA EP y cargarlo en la memoria del computador bajo el nombre de df (un DataFrame o estructura de tabla). Esto nos permite empezar a explorar, limpiar y manipular las variables de forma estructurada en Python.

# 2. Análisis Descriptivo & Estadística Descriptiva
Identificamos la estructura general y las propiedades estadísticas de las variables numéricas.

Aquí se utiliza el método head() de Pandas para imprimir en pantalla las primeras 5 filas de la tabla de datos (df). Esto sirve para que se valide visualmente que los datos se cargaron correctamente, revisando los nombres de las columnas de los procesos de contratacion y el tipo de valores que contiene cada una.

In [ ]:
# Visualizamos las primeras 5 filas para entender la estructura inicial
print("--- Primeros Registros ---")
print(df.head())

--- Primeros Registros ---
  Código del proceso Fecha de Adjudicación Metodo de Contratacion   \
0  CE-20220002242374            20/07/2022                     NaN   
1  CE-20220002242377            20/07/2022                     NaN   
2  CE-20220002242373            20/07/2022                     NaN   
3  CE-20220002242371            20/07/2022                     NaN   
4  CE-20220002242375            20/07/2022                     NaN   

   Tipo de contratación Fecha de Publicacion Código del proceso.1  \
0  Catálogo Electrónico           20/07/2022    CE-20220002242374   
1  Catálogo Electrónico           20/07/2022    CE-20220002242377   
2  Catálogo Electrónico           20/07/2022    CE-20220002242373   
3  Catálogo Electrónico           20/07/2022    CE-20220002242371   
4  Catálogo Electrónico           20/07/2022    CE-20220002242375   

                                              Objeto   Monto de Adjudicación  \
0  *NEUMATICO 205/75R15 98/95Q TODA POSICION APLI...     

In [ ]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89183 entries, 0 to 89182
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Código del proceso       89183 non-null  object 
 1   Fecha de Adjudicación    89183 non-null  object 
 2   Metodo de Contratacion   88789 non-null  object 
 3   Tipo de contratación     89183 non-null  object 
 4   Fecha de Publicacion     14386 non-null  object 
 5   Código del proceso.1     83518 non-null  object 
 6   Objeto                   83518 non-null  object 
 7    Monto de Adjudicación   81833 non-null  float64
 8   RUC del Oferente         89183 non-null  object 
 9   RUC de la entidad        89183 non-null  object 
 10  Entidad                  89183 non-null  object 
 11  Ámbito de Control        87916 non-null  object 
 12  Índice                   89183 non-null  object 
 13  Competencia              88789 non-null  float64
 14  Trazabilidad          

# Interpretación df.info()
El resultado del método df.info() revela que el dataset está estructurado como un DataFrame de Pandas que contiene exactamente 89,183 registros (indexados del 0 al 89,182) y 16 columnas en total. Técnicamente, se observa que no todas la variable tienen 89,183 registros, es decir, variables como: "metodo de contratación" tiene 88,789 registros; "fecha de publicación" tiene 14386 registros, lo que confirma la presencia total de datos faltantes en esta etapa. Respecto a la naturaleza de las variables, 12 de ellas corresponden a información  almacenadas como (object), mientras que la variable objetivo original (Indice) está registrada como cadenas de texto y numeros (object), demandando todo el conjunto un consumo de memoria RAM sumamente ligero de apenas 11.6 MB.

# Eliminacion de los datos definidos como "No determinados" de la Variable Objetivo "Indice"
A continuacion eliminaramos los datos de la variable "Índice" que esten definidos como "no determinado", ya que no aportan nada a nuestro analisis, no existe evidencia de que tengan un mayor riesgo.

In [ ]:
#Eliminar del dataset df, de la variable objetivo "Indice" el valor como "No determinado "
df_modelo = df[df["Índice"] != "No determinado"].copy()
print(df_modelo.info())

<class 'pandas.core.frame.DataFrame'>
Index: 89030 entries, 75 to 89182
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Código del proceso       89030 non-null  object 
 1   Fecha de Adjudicación    89030 non-null  object 
 2   Metodo de Contratacion   88789 non-null  object 
 3   Tipo de contratación     89030 non-null  object 
 4   Fecha de Publicacion     14233 non-null  object 
 5   Código del proceso.1     83365 non-null  object 
 6   Objeto                   83365 non-null  object 
 7    Monto de Adjudicación   81680 non-null  float64
 8   RUC del Oferente         89030 non-null  object 
 9   RUC de la entidad        89030 non-null  object 
 10  Entidad                  89030 non-null  object 
 11  Ámbito de Control        87763 non-null  object 
 12  Índice                   89030 non-null  object 
 13  Competencia              88789 non-null  float64
 14  Trazabilidad             8

In [ ]:
#Convertimos la variable"Índice" de (object) a numerica decimal (float)
df_modelo["Índice"] = pd.to_numeric(
df_modelo["Índice"],
errors="coerce"
)
df_modelo.reset_index(drop=True, inplace=True) # --- Reiniciamos el Indice del DataFrame df_modelo
print(df_modelo.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89030 entries, 0 to 89029
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Código del proceso       89030 non-null  object 
 1   Fecha de Adjudicación    89030 non-null  object 
 2   Metodo de Contratacion   88789 non-null  object 
 3   Tipo de contratación     89030 non-null  object 
 4   Fecha de Publicacion     14233 non-null  object 
 5   Código del proceso.1     83365 non-null  object 
 6   Objeto                   83365 non-null  object 
 7    Monto de Adjudicación   81680 non-null  float64
 8   RUC del Oferente         89030 non-null  object 
 9   RUC de la entidad        89030 non-null  object 
 10  Entidad                  89030 non-null  object 
 11  Ámbito de Control        87763 non-null  object 
 12  Índice                   89030 non-null  float64
 13  Competencia              88789 non-null  float64
 14  Trazabilidad          

# Interpretación
Ahora hemos creado un nuevo dataframe llamado df_modelo, en el cual, hemos eliminado los registros correspondientes a "No determinado" de la variable "Indice", con ello, tenemos 89030 registros en total, ademas hemos cambiado el tipo de variable de (object) a (float), y por ultimo se ha reseteado el indice de nuestro dataframe

# Normalizacion de Variables
A continuacion vamos a normalizar nuestras variables, monto de adjudicación e indice

In [ ]:
# Normalizar monto
df_modelo["Monto_Norm"] = scaler.fit_transform(
df_modelo[[" Monto de Adjudicación"]]
)

# Normalizar índice
df_modelo["Indice_Norm"] = scaler.fit_transform(
df_modelo[["Índice"]]
)


In [ ]:
print(df_modelo.head())

  Código del proceso Fecha de Adjudicación Metodo de Contratacion   \
0  CE-20220002317888            15/12/2022                     NaN   
1  CE-20220002283689            12/10/2022                     NaN   
2  CE-20220002283689            12/10/2022                     NaN   
3  CE-20220002283689            12/10/2022                     NaN   
4  CE-20220002283689            12/10/2022                     NaN   

   Tipo de contratación Fecha de Publicacion Código del proceso.1  \
0  Catálogo Electrónico           15/12/2022    CE-20220002317888   
1  Catálogo Electrónico           12/10/2022    CE-20220002283689   
2  Catálogo Electrónico           12/10/2022    CE-20220002283689   
3  Catálogo Electrónico           12/10/2022    CE-20220002283689   
4  Catálogo Electrónico           12/10/2022    CE-20220002283689   

                                              Objeto   Monto de Adjudicación  \
0               *CAMIONETA DOBLE CABINA 4X4 MODELO 1            7.247240e+05   
1  P

#  Interpretación
¿Porque es importante normalizar estas variables? Esto nos permite, que las variables esten en escalas iguales para compararlas. Como se evidencia se han creado dos variables dentro de nuestro dataframe "Monto_Norm" e "Indice_Norm", estan normalizadas en una escala del 0 al 1, lo cual las hace comparables

# Creamos Dataframe solo con los procesos de contratacion de ETAPA EP
A continuacion se va a crear otro dataframe solo con los procesos de ETAPA EP, ya que el resto de procesos no las llevo a cabo la empresa y por lo tanto, no tiene sentido analizarlos.

In [ ]:
df_etapa = df_modelo[
df_modelo["Entidad"].str.upper() ==
"EMPRESA PÚBLICA MUNICIPAL DE TELECOMUNICACIONES, AGUA POTABLE, ALCANTARILLADO Y SANEAMIENTO DE CUENCA ETAPA EP"
].copy()

In [ ]:
#Verificamos la data de etapa
df_etapa.info()

<class 'pandas.core.frame.DataFrame'>
Index: 928 entries, 0 to 88174
Data columns (total 19 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Código del proceso       928 non-null    object 
 1   Fecha de Adjudicación    928 non-null    object 
 2   Metodo de Contratacion   687 non-null    object 
 3   Tipo de contratación     928 non-null    object 
 4   Fecha de Publicacion     675 non-null    object 
 5   Código del proceso.1     805 non-null    object 
 6   Objeto                   805 non-null    object 
 7    Monto de Adjudicación   774 non-null    float64
 8   RUC del Oferente         928 non-null    object 
 9   RUC de la entidad        928 non-null    object 
 10  Entidad                  928 non-null    object 
 11  Ámbito de Control        928 non-null    object 
 12  Índice                   928 non-null    float64
 13  Competencia              687 non-null    float64
 14  Trazabilidad             687 

# Interpretación
Ahora tenemos un nuevo Dataframe con el nombre "df_etapa", tiene 928 registros y 19 columnas, con 11 variables (object) y 8 variables con numeros decimales (float)

# Construir el Score de Priorizacion de Procesos de Contratación
Conocemos que el indice corresponde al promedio de: competencia, trazabilidad, transparencia, Temporabilidad, por lo tanto nuestro analisis sera en base solo al Índice ya que incluye el resto de indices nombrados, ademas, sabemos que si se acercan a uno, mejor sera el score de calificación, en ese caso vamos  invertir para que nos diga, si está mas bajo, existe mas riesgo de auditoria. Luego se establecera los siguientes criterios y pesos: Índice 70% Monto 30%.

In [ ]:
#Invertimos
df_etapa["Indice_Riesgo"] = 1 - df_etapa["Indice_Norm"]
df_etapa["Monto_Riesgo"] = 1 - df_modelo["Monto_Norm"]

#Calculamos el Score de Auditoría
df_etapa["Score_Auditoria"] = (
0.70 * df_etapa["Indice_Riesgo"]
+ 0.30 * df_etapa["Monto_Riesgo"])

# Ranking de procesos de auditoria
A continuacion se establecerá el ranking de procesos a auditar para el PAC 2027 de ETAPA EP

In [ ]:
ranking_auditoria = (
df_etapa
.sort_values(
by="Score_Auditoria",
ascending=False
)
)

# TOP 20 Procesos de Contratacion
Establaceremos y Observaremos, cuales son los 20 procesos mas riesgosos que se auditaran en base al score (criterios dados)

In [ ]:
ranking_auditoria[
[
"Código del proceso",
"Objeto",
" Monto de Adjudicación",
"Índice",
"Score_Auditoria"
]
].head(20)

,Código del proceso,Objeto,Monto de Adjudicación,Índice,Score_Auditoria
70,SIE-ETAPA-31-2022,ADQUISICIÓN DE TUBERÍA Y ACCESORIOS DE PVC PAR...,289904.78125,0.763462,0.999250
62559,CE-20260003035122-312203,Orden de compra para adquirir los siguientes p...,15.42000,0.786000,0.933301
62560,CE-20260003035123-312203,Orden de compra para adquirir los siguientes p...,83.55000,0.786000,0.933301
741,CE-20250002889438-312203,Orden de compra para adquirir los siguientes p...,98.90000,0.786000,0.933301
478,CE-20240002764988-312203,Orden de compra para adquirir los siguientes p...,107.41000,0.786000,0.933301
647,CE-20250002909409-312203,Orden de compra para adquirir los siguientes p...,107.41000,0.786000,0.933301
62428,CE-20260002989877-312203,Orden de compra para adquirir los siguientes p...,130.80000,0.786000,0.933300
62430,CE-20260002989954-312203,Orden de compra para adquirir los siguientes p...,157.78000,0.786000,0.933300
485,CE-20240002765020-312203,Orden de compra para adquirir los siguientes p...,186.99000,0.786000,0.933300
486,CE-20240002765032-312203,Orden de compra para adquirir los siguientes p...,215.04000,0.786000,0.933300


# Clasificación del PAC
Ahora vamos a clasificar y asignar el nivel de prioridad de los procesos de contratacion del dataframe de ETAPA EP (df_etapa)

In [ ]:
q80 = df_etapa["Score_Auditoria"].quantile(0.80)
q95 = df_etapa["Score_Auditoria"].quantile(0.95)

df_etapa["Nivel_Prioridad"] = np.where(
df_etapa["Score_Auditoria"] >= q95,
"CRITICA",
np.where(
df_etapa["Score_Auditoria"] >= q80,
"ALTA",
"NORMAL"
)
)

# Interpretación
Al clasificarlos por q80 y q95 estamos haciendo los siguiente:

q95 = valor por encima del cual se encuentra el 5% de los procesos con mayor score, el cual se considera procesos con riesgo critico para realizar una auditoria.

q80 = valor por encima del cual se encuentra el 20% de los procesos con mayor score, el cual se considera procesos con riesgo alto para realizar una auditoria.

Los procesos de contraccion que esten por debajo de este 20% se consideran con riesgo normal, para realizar una auditoría


# Seleccionar procesos de Auditoria
Finalmente seleccionamos los procesos que segun su score de calificacion, se priorizaran para una revisión del PAC

In [ ]:
procesos_pac = df_etapa[
df_etapa["Nivel_Prioridad"].isin(
["CRITICA", "ALTA", "NORMAL"]
)
]
procesos_pac.head()

,Código del proceso,Fecha de Adjudicación,Metodo de Contratacion,Tipo de contratación,Fecha de Publicacion,Código del proceso.1,Objeto,Monto de Adjudicación,RUC del Oferente,RUC de la entidad,...,Competencia,Trazabilidad,Transparencia,Temporalidad,Monto_Norm,Indice_Norm,Indice_Riesgo,Monto_Riesgo,Score_Auditoria,Nivel_Prioridad
0,CE-20220002317888,15/12/2022,NaN,Catálogo Electrónico,15/12/2022,CE-20220002317888,*CAMIONETA DOBLE CABINA 4X4 MODELO 1,7.247240e+05,1891748376001,0160050020001,...,NaN,NaN,NaN,NaN,0.006248,1.0,0.0,0.993752,0.298126,NORMAL
1,CE-20220002283689,12/10/2022,NaN,Catálogo Electrónico,12/10/2022,CE-20220002283689,PUNTO DE SERVICIO INSTITUCIONAL DE 24 HORAS DE...,1.657268e+06,0992723211001,0160050020001,...,NaN,NaN,NaN,NaN,0.014288,1.0,0.0,0.985712,0.295714,NORMAL
2,CE-20220002283689,12/10/2022,NaN,Catálogo Electrónico,12/10/2022,CE-20220002283689,PUNTO DE SERVICIO INSTITUCIONAL DE 12 HORAS - ...,5.351071e+05,0992723211001,0160050020001,...,NaN,NaN,NaN,NaN,0.004613,1.0,0.0,0.995387,0.298616,NORMAL
3,CE-20220002283689,12/10/2022,NaN,Catálogo Electrónico,12/10/2022,CE-20220002283689,PUNTO SERVICIO INSTITUCIONAL DE 12 HORAS - LUN...,9.994223e+04,0992723211001,0160050020001,...,NaN,NaN,NaN,NaN,0.000862,1.0,0.0,0.999138,0.299742,NORMAL
4,CE-20220002283689,12/10/2022,NaN,Catálogo Electrónico,12/10/2022,CE-20220002283689,PUNTO DE SERVICIO INSTITUCIONAL DE 8 HORAS - L...,2.177379e+04,0992723211001,0160050020001,...,NaN,NaN,NaN,NaN,0.000188,1.0,0.0,0.999812,0.299944,NORMAL


# Exportar el analisis a Excel
A continuacion exportamos la informacion a Excel

In [ ]:
procesos_pac.sort_values(
by="Score_Auditoria",
ascending=False
).to_excel(
"Procesos_PAC_Seleccionados.xlsx",
index=False
)
#Descargar archivo en la PC
from google.colab import files

files.download("Procesos_PAC_Seleccionados.xlsx")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>